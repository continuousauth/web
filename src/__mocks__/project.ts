import { FullProject } from '../common/types';

export const mockProject = (): FullProject => ({
  secret: 'my_secret',
  enabled: true,
  requester_circleCI: null,
  requester_gitHub: null,
  responder_slack: null,
  id: '123',
  repoName: 'my-repo',
  repoOwner: 'my-owner',
  defaultBranch: 'main',
});
