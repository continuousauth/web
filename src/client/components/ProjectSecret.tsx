import * as React from 'react';
import { Button, Heading, Pane, TextInput } from 'evergreen-ui';

import { FullProject } from '../../common/types';

import styles from './ProjectSecret.scss';

export interface Props {
  project: FullProject;
}

export function ProjectSecret({ project }: Props) {
  const [isVisible, setVisisble] = React.useState(false);

  return (
    <Pane>
      <Heading>Project Secret</Heading>
      <Pane className={styles.inputContainer}>
        <TextInput
          className={styles.input}
          value={isVisible ? project.secret : '••••••••••••••••••••••••••••••••••••••••••••••••••'}
          disabled
        />
        <Button className={styles.button} intent="danger" onClick={() => setVisisble(!isVisible)}>
          {isVisible ? 'Hide' : 'Show'}
        </Button>
      </Pane>
    </Pane>
  );
}
