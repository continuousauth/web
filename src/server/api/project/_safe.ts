import * as crypto from 'crypto';
import * as express from 'express';

import { Project } from '../../db/models';
import { ExpressRequest, hasAdminAccessToTargetRepo } from '../../helpers/_middleware';

export const generateNewSecret = (size: number) => {
  if (size > 256) throw new Error('size must be <= 256');
  return crypto.randomBytes(256).toString('hex').slice(0, size);
};

export const sanitizeProject = (project: Project) => project;

export const getProjectFromIdAndCheckPermissions = async (
  id: string,
  req: ExpressRequest,
  res: express.Response,
) => {
  const project = await Project.findOne({
    where: {
      id: id,
      enabled: true,
    },
    include: Project.allIncludes,
  });
  // Project not existing and user not having permission should look identical
  // so that folks can't sniff which projects are active on CFA
  if (!project) {
    res.status(401).json({
      error: 'Current user is not an admin of the target repository',
    });
    return null;
  }

  if (!(await hasAdminAccessToTargetRepo(req, project.id))) {
    res.status(401).json({
      error: 'Current user is not an admin of the target repository',
    });
    return null;
  }

  return project;
};
