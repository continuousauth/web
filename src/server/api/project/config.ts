import axios from 'axios';
import * as debug from 'debug';
import * as express from 'express';
import * as Joi from 'joi';

import { validate, hasAdminAccessToTargetRepo } from '../../helpers/_middleware';
import { createA } from '../../helpers/a';
import {
  Project,
  SlackResponderLinker,
  withTransaction,
  CircleCIRequesterConfig,
  TravisCIRequesterConfig,
} from '../../db/models';
import { getProjectFromIdAndCheckPermissions } from './_safe';

const d = debug('cfa:server:api:project:config');
const a = createA(d);

export async function updateCircleEnvVars(project: Project, accessToken: string) {
  const client = axios.create({
    baseURL: 'https://circleci.com/api/v1.1',
    auth: {
      username: accessToken,
      password: '',
    },
    validateStatus: () => true,
  });

  const existing = await client.get(
    `/project/github/${project.repoOwner}/${project.repoName}/envvar`,
  );
  if (existing.status !== 200) return;

  if (existing.data.find(item => item.name === 'CFA_SECRET')) {
    await client.delete(
      `/project/github/${project.repoOwner}/${project.repoName}/envvar/CFA_SECRET`,
    );
  }
  await client.post(`/project/github/${project.repoOwner}/${project.repoName}/envvar`, {
    name: 'CFA_SECRET',
    value: project.secret,
  });

  if (existing.data.find(item => item.name === 'CFA_PROJECT_ID')) {
    await client.delete(
      `/project/github/${project.repoOwner}/${project.repoName}/envvar/CFA_PROJECT_ID`,
    );
  }
  await client.post(`/project/github/${project.repoOwner}/${project.repoName}/envvar`, {
    name: 'CFA_PROJECT_ID',
    value: project.id,
  });
}

export function configRoutes() {
  const router = express();

  router.post(
    '/:id/config/requesters/circleci',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
        body: {
          accessToken: Joi.string()
            .min(1)
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        const response = await axios.get(
          `https://circleci.com/api/v1.1/project/gh/${project.repoOwner}/${project.repoName}/checkout-key`,
          {
            auth: {
              username: req.body.accessToken,
              password: '',
            },
            validateStatus: () => true,
          },
        );

        if (response.status !== 200) {
          return res.status(401).json({
            error:
              'That token is not valid for the current project, or the repository is not configured on CircleCI',
          });
        }

        await updateCircleEnvVars(project, req.body.accessToken);

        const newProject = await withTransaction(async t => {
          const config = await CircleCIRequesterConfig.create(
            {
              accessToken: req.body.accessToken,
            },
            {
              returning: true,
            },
          );
          await project.resetAllRequesters(t);
          project.requester_circleCI_id = config.id;
          await project.save({ transaction: t });
          return await Project.findByPk(project.id, {
            include: Project.allIncludes,
            transaction: t,
          });
        });

        res.json(newProject);
      },
    ),
  );

  router.post(
    '/:id/config/requesters/travisci',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
        body: {
          accessToken: Joi.string()
            .min(1)
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        const response = await axios.get(
          `https://api.travis-ci.org/repo/${project.repoOwner}%2F${project.repoName}`,
          {
            headers: {
              'Travis-API-Version': '3',
              Authorization: `token ${req.body.accessToken}`,
            },
            validateStatus: () => true,
          },
        );

        if (response.status !== 200) {
          return res.status(401).json({
            error:
              'That token is not valid for the current project, or the repository is not configured on Travis CI',
          });
        }

        const newProject = await withTransaction(async t => {
          const config = await TravisCIRequesterConfig.create(
            {
              accessToken: req.body.accessToken,
            },
            {
              returning: true,
            },
          );
          await project.resetAllRequesters(t);
          project.requester_travisCI_id = config.id;
          await project.save({ transaction: t });
          return await Project.findByPk(project.id, {
            include: Project.allIncludes,
            transaction: t,
          });
        });

        res.json(newProject);
      },
    ),
  );

  router.post(
    '/:id/config/responders/slack',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        const linker = await withTransaction(async t => {
          await SlackResponderLinker.destroy({
            where: {
              projectId: project.id,
            },
            transaction: t,
          });
          return await SlackResponderLinker.create(
            {
              projectId: project.id,
            },
            {
              transaction: t,
              returning: true,
            },
          );
        });

        res.json({
          linker,
          slackClientId: process.env.SLACK_CLIENT_ID,
        });
      },
    ),
  );

  router.patch(
    '/:id/config/responders/slack',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
        body: {
          usernameToMention: Joi.string()
            .min(1)
            .max(50)
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        if (!project.responder_slack) {
          return res.status(400).json({
            error: 'Project is not configured to use Slack as a responder',
          });
        }

        project.responder_slack.usernameToMention = req.body.usernameToMention;
        await project.responder_slack.save();

        res.json(project);
      },
    ),
  );

  return router;
}
