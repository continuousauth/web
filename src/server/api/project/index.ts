import * as debug from 'debug';
import * as express from 'express';
import * as Joi from 'joi';
import * as Octokit from '@octokit/rest';

import { createA } from '../../helpers/a';
import { validate, hasAdminAccessToTargetRepo } from '../../helpers/_middleware';
import { Project, withTransaction, OTPRequest } from '../../db/models';
import { configRoutes } from './config';
import { sanitizeProject, generateNewSecret, getProjectFromIdAndCheckPermissions } from './_safe';

const d = debug('cfa:server:api:auth');
const a = createA(d);

export function projectRoutes() {
  const router = express();

  router.post(
    '/',
    validate(
      {
        a,
        body: {
          repoId: Joi.number()
            .integer()
            .required(),
        },
      },
      async (req, res) => {
        if (!(await hasAdminAccessToTargetRepo(req, `${req.body.repoId}`))) {
          return res.status(401).json({
            error: 'Current user is not an admin of the target repository',
          });
        }

        const github = new Octokit({
          auth: req.user.accessToken,
        });

        const repoResponse = await github.request('GET /repositories/:id', {
          id: req.body.repoId,
        });
        const repo: Octokit.ReposGetResponse = repoResponse.data;

        const existingProject = await Project.findByPk(req.body.repoId);
        if (existingProject) {
          existingProject.repoName = repo.name;
          existingProject.repoOwner = repo.owner.login;
          // If the project existed but was disabled let's enable it and
          // generate a new secret so that the old info is invalid.
          if (!existingProject.enabled) {
            existingProject.enabled = true;
            existingProject.secret = generateNewSecret(256);
          }
          await existingProject.save();
          return res.status(200).json(existingProject);
        }

        const project = await Project.create({
          id: req.body.repoId,
          repoName: repo.name,
          repoOwner: repo.owner.login,
          secret: generateNewSecret(256),
          defaultBranch: repo.default_branch,
        });
        res.status(201).json(project);
      },
    ),
  );

  router.get(
    '/:id',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
      },
      async (req, res) => {
        const project = await Project.findOne({
          where: {
            id: req.params.id,
            enabled: true,
          },
          include: Project.allIncludes,
        });
        // Project not existing and user not having permission should look identical
        // so that folks can't sniff which projects are active on CFA
        if (!project) {
          return res.status(401).json({
            error: 'Current user is not an admin of the target repository',
          });
        }

        if (!(await hasAdminAccessToTargetRepo(req, project.id))) {
          return res.status(401).json({
            error: 'Current user is not an admin of the target repository',
          });
        }

        res.json(sanitizeProject(project));
      },
    ),
  );

  router.get(
    '/:id/log',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        res.json(
          (
            await OTPRequest.findAll({
              where: {
                projectId: project.id,
              },
            })
          )
            .map(req => {
              const simpleReq = req.get() as OTPRequest;
              delete simpleReq.proof;
              delete simpleReq.requestMetadata;
              delete simpleReq.responseMetadata;
              return simpleReq;
            })
            .sort((a, b) => b.requested.getTime() - a.requested.getTime()),
        );
      },
    ),
  );

  router.patch(
    '/:id/secret',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        project.secret = generateNewSecret(256);
        await project.save();

        res.json(sanitizeProject(project));
      },
    ),
  );

  router.delete(
    '/:id',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        await withTransaction(async t => {
          project.enabled = false;
          await project.resetAllResponders(t);
          await project.resetAllRequesters(t);
          await project.save({ transaction: t });
        });

        res.json(sanitizeProject(project));
      },
    ),
  );

  router.use(configRoutes());

  return router;
}
