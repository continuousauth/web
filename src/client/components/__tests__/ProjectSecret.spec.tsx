/**
 * @jest-environment jsdom
 */

import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { ProjectSecret } from '../ProjectSecret';

import { mockProject } from '../../../__mocks__/project';

describe('<ProjectSecret />', () => {
  it('Should render a disabled input', () => {
    const wrapper = shallow(<ProjectSecret project={mockProject()} />);
    expect(wrapper).toMatchSnapshot();
    const mounted = mount(<ProjectSecret project={mockProject()} />);
    expect(mounted.find('ForwardRef(Button)')).toHaveLength(1);
  });

  it('should initially dispplay asterix (not the secret)', () => {
    const mounted = mount(<ProjectSecret project={mockProject()} />);
    expect(
      mounted
        .find('ForwardRef(Button)')
        .children()
        .text(),
    ).toMatchInlineSnapshot(`"Show"`);
    expect(mounted.find('ForwardRef(TextInput)').prop('value')).toMatchInlineSnapshot(
      `"••••••••••••••••••••••••••••••••••••••••••••••••••"`,
    );
  });

  it('should toggle to display the secret when the button is clicked', () => {
    const mounted = mount(<ProjectSecret project={mockProject()} />);
    mounted.find('ForwardRef(Button)').simulate('click');
    expect(
      mounted
        .find('ForwardRef(Button)')
        .children()
        .text(),
    ).toMatchInlineSnapshot(`"Hide"`);
    expect(mounted.find('ForwardRef(TextInput)').prop('value')).toBe(mockProject().secret);
  });
});
