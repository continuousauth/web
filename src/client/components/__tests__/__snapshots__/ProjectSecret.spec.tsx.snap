// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectSecret /> Should render a disabled input 1`] = `
<Memo(ForwardRef(Pane))>
  <Memo(ForwardRef(Heading))>
    Project Secret
  </Memo(ForwardRef(Heading))>
  <Memo(ForwardRef(Pane))
    className="inputContainer"
  >
    <Memo(ForwardRef(TextInput))
      className="input"
      disabled={true}
      value="••••••••••••••••••••••••••••••••••••••••••••••••••"
    />
    <Memo(ForwardRef(Button))
      className="button"
      intent="danger"
      onClick={[Function]}
    >
      Show
    </Memo(ForwardRef(Button))>
  </Memo(ForwardRef(Pane))>
</Memo(ForwardRef(Pane))>
`;
